File: dsa_leetcode_beginner/0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

class Solution {
    buildTree(preorder, inorder) {
        // Create a map to efficiently find the index of any element in inorder list
        const io_map = new Map();
        for (let i = 0; i < inorder.length; i++) {
            io_map.set(inorder[i], i);
        }
        
        // Call recursive helper function to build the tree
        return this.splitTree(preorder, io_map, 0, 0, inorder.length - 1);
    }

    splitTree(preorder, io_map, rootIndex, left, right) {
        // Base case: if the left index exceeds the right index, subtree is empty
        if (left > right) {
            return null;
        }
        
        // Create the root node with the current root element
        const root = new TreeNode(preorder[rootIndex]);
        
        // Find the index of the root element in inorder list
        const mid = io_map.get(preorder[rootIndex]);
        
        // Recursively build the left subtree
        if (mid > left) {
            root.left = this.splitTree(preorder, io_map, rootIndex + 1, left, mid - 1);
        }
        
        // Recursively build the right subtree
        if (mid < right) {
            root.right = this.splitTree(preorder, io_map, rootIndex + mid - left + 1, mid + 1, right);
        }
        
        return root;
    }
}

// Example usage
// const solution = new Solution();
// const tree = solution.buildTree([3, 9, 20, 15, 7], [9, 3, 15, 20, 7]);

File: dsa_leetcode_beginner/0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
#include <unordered_map>
#include <vector>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    TreeNode* buildTree(std::vector<int>& preorder, std::vector<int>& inorder) {
        // Create a map to efficiently find the index of any element in inorder list
        std::unordered_map<int, int> io_map;
        for (int i = 0; i < inorder.size(); ++i) {
            io_map[inorder[i]] = i;
        }
        
        // Call recursive helper function to build the tree
        return splitTree(preorder, io_map, 0, 0, inorder.size() - 1);
    }

private:
    TreeNode* splitTree(const std::vector<int>& preorder, const std::unordered_map<int, int>& io_map, int rootIndex, int left, int right) {
        // Base case: if the left index exceeds the right index, subtree is empty
        if (left > right) {
            return nullptr;
        }
        
        // Create the root node with the current root element
        TreeNode* root = new TreeNode(preorder[rootIndex]);
        
        // Find the index of the root element in inorder list
        int mid = io_map.at(preorder[rootIndex]);
        
        // Recursively build the left subtree
        if (mid > left) {
            root->left = splitTree(preorder, io_map, rootIndex + 1, left, mid - 1);
        }
        
        // Recursively build the right subtree
        if (mid < right) {
            root->right = splitTree(preorder, io_map, rootIndex + mid - left + 1, mid + 1, right);
        }
        
        return root;
    }
};
